Add dstack pool

TheBits opened the pull request:
This PR adds pool support.
Technical requirements to pools recorded in the #790.

Commit:
pool initial

Commit:
pool add

Commit:
merge pool logic

Commit:
add local instance to the pool

Commit:
multitask

Commit:
update

Commit:
update

Commit:
update

Commit:
debug

Commit:
debug

Commit:
fix mypy

Commit:
improve dstack run

Commit:
fixup! improve dstack run

Commit:
Merge remote-tracking branch 'origin/master' into pool

Egor-S committed:
Refactor check_relevance to use gpuhunt

Egor-S committed:
Fix tests discovery for Python 3.8

Commit:
Improve profile.pool_name handling

Commit:
Merge remote-tracking branch 'origin/master' into pool

Commit:
Merge remote-tracking branch 'origin/pool-gpuhunt' into pool

Egor-S committed:
Fix spot policy mapping

Commit:
fixup! Improve profile.pool_name handling

Commit:
Merge remote-tracking branch 'origin/pool-gpuhunt' into pool

Commit:
Merge remote-tracking branch 'origin/master' into pool

Commit:
Merge remote-tracking branch 'origin/master' into pool

Commit:
small fixies

Egor-S committed:
Fix rich formatting, require --name

Egor-S committed:
Filter out deleted instances in pools

Commit:
Update runner, fix status

Commit:
Merge remote-tracking branch 'origin/pool-cli-improvements' into pool

Commit:
Merge remote-tracking branch 'origin/master' into pool

Commit:
Improve job statuses

Commit:
Merge remote-tracking branch 'origin/master' into pool

Commit:
fix hardcode

Commit:
Fix CI

r4victor commented on src/dstack/_internal/cli/commands/pool.py:
```diff
@@ -0,0 +1,396 @@
+logger = get_logger(__name__)
+
+
+class PoolCommand(APIBaseCommand):  # type: ignore[misc]
```
Remove `# type: ignore` comments since we're not enforcing type checking. Type checking should be a separate issue if to be introduced.

TheBits replied:
Deleted all `# type: ignore` comment.

r4victor commented on src/dstack/_internal/cli/main.py:
```diff
@@ -6,11 +6,12 @@
 from dstack._internal.cli.commands.server import ServerCommand
 from dstack._internal.cli.commands.stop import StopCommand
-from dstack._internal.cli.utils.common import colors, console
+from dstack._internal.cli.utils.common import _colors, console
```
Change _colors to colors or COLORS. Leading underscore means the var is private to the module.

Egor-S replied:
I've changed it recently to private because we were tempted to use it for rich formatting with the theme-customized `console`. This import is the only place, where `_colors` is being used "externally"

TheBits replied:
I'm going to leave this part unchanged.

r4victor commented on src/dstack/_internal/core/backends/aws/compute.py:
```diff
@@ -94,6 +97,64 @@ def terminate_instance(
             else:
                 raise e
 
+    def create_instance(
```
Logic is duplicated in `create_instance()` and `run_job()`. Should be refactored to avoid duplication.

TheBits replied:
I've removed logic duplicates

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -2,25 +2,50 @@
 logger = get_logger(__name__)
 
 
+class SSHKeys(BaseModel):
```
This is no place for common models. Our approach is to place models in dstack/src/dstack/_internal/core/models. E.g. pool.py.

r4victor replied:
SSHKeys should be renamed SSHKey or SSHKeyPair

TheBits replied:
I've moved class to the dstack/src/dstack/_internal/core/models/instances.py and renamed to SSHKey

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -39,10 +64,20 @@ def run_job(
     ) -> LaunchedInstanceInfo:
         pass
 
+    @abstractmethod
```
abstractmethod should be removed since it forces all backends to implement create_instance().

TheBits replied:
Thanks! I've removed abstractmethod.

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -115,12 +150,16 @@ def get_cloud_config(**config) -> str:
+    #     bucket = "dstack-runner-downloads"
+    # url =f"https://{bucket}.s3.eu-west-1.amazonaws.com/{build}/binaries/dstack-shim-linux-amd64"
+
+    url = "https://da344481-89d9-4f32-bd6a-8e0b47b1eb8c.selstorage.ru/dstack-shim"
```
Fix

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -183,13 +222,20 @@ def get_docker_commands(authorized_keys: List[str]) -> List[str]:
+    #     bucket = "dstack-runner-downloads"
+    # url = f'https://{bucket}.s3.eu-west-1.amazonaws.com/{build}/binaries/dstack-runner-linux-amd64'
+
+    url = "https://da344481-89d9-4f32-bd6a-8e0b47b1eb8c.selstorage.ru/dstack-runner"
```
Fix

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -257,3 +303,102 @@ def get_dstack_gateway_commands() -> List[str]:
+    #     bucket = "dstack-runner-downloads"
+    # url = f"https://{bucket}.s3.eu-west-1.amazonaws.com/{build}/binaries/dstack-shim-linux-amd64"
+
+    url = "https://da344481-89d9-4f32-bd6a-8e0b47b1eb8c.selstorage.ru/dstack-shim"
```
Fix

TheBits replied:
Fixed.

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -257,3 +303,102 @@ def get_dstack_gateway_commands() -> List[str]:
+    #     bucket = "dstack-runner-downloads"
+    # url = f"https://{bucket}.s3.eu-west-1.amazonaws.com/{build}/binaries/dstack-runner-linux-amd64"
+
+    url = "https://da344481-89d9-4f32-bd6a-8e0b47b1eb8c.selstorage.ru/dstack-runner"
```
Fix

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/core/backends/gcp/compute.py:
```diff
@@ -70,14 +73,80 @@ def get_offers(
         except google.api_core.exceptions.NotFound:
             pass
 
+    def create_instance(
```
As with AWS, logic is duplicated in `create_instance()` and `run_job()`. Should be refactored to avoid duplication.

TheBits replied:
I've removed logic duplication.

r4victor commented on src/dstack/_internal/server/background/tasks/process_pools.py:
```diff
@@ -0,0 +1,181 @@
+            current_time = get_current_datetime().replace(tzinfo=datetime.timezone.utc)
+
+            delta = datetime.timedelta(
+                seconds=parse_pretty_duration(instance.termination_idle_time)
```
termination_idle_time should not be parsed at the time of processing. The db should should store already parsed termination_idle_time

TheBits replied:
Yes, I fixed it

r4victor commented on src/dstack/_internal/server/background/tasks/process_submitted_jobs.py:
```diff
@@ -74,21 +86,111 @@ async def _process_submitted_job(session: AsyncSession, job_model: JobModel):
+        pool_instances, profile, run_spec.configuration.resources, status=InstanceStatus.READY
+    )
+
+    logger.info(*job_log(f"num relevance {len(relevant_instances)}", job_model))
```
Remove unnecessary logging or make it clear

TheBits replied:
Removed

r4victor commented on src/dstack/_internal/server/routers/runs.py:
```diff
@@ -58,9 +66,48 @@ async def get_run(
+    body: GetOffersRequest,
+    session: AsyncSession = Depends(get_session),
+    user_project: Tuple[UserModel, ProjectModel] = Depends(ProjectMember()),
+) -> Tuple[str, List[InstanceOfferWithAvailability]]:
```
Define models for complex responses since working with tuples at the API level is inconvenient (e.g. hard-to-read openapi schema).

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/server/background/tasks/process_submitted_jobs.py:
```diff
@@ -74,21 +86,111 @@ async def _process_submitted_job(session: AsyncSession, job_model: JobModel):
+            if pool_item.name == DEFAULT_POOL_NAME:
+                pool = pool_item
+        if pool is None:
+            pool = PoolModel(
```
Refactor pool creation. You already have `create_pool_model()`.

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/server/background/tasks/process_submitted_jobs.py:
```diff
@@ -74,21 +86,111 @@ async def _process_submitted_job(session: AsyncSession, job_model: JobModel):
+        sorted_instances = sorted(relevant_instances, key=lambda instance: instance.name)
+        instance = sorted_instances[0]
+
+        # need lock
```
I consider locking a critical functionality here since now job can be assigned to instance that would be terminated by `process_pools`.

r4victor commented on src/dstack/_internal/server/background/tasks/process_pools.py:
I insist on refactoring `process_pool.py` to multiple properly named files: separately for processing detached instances for termination and checking shim.

r4victor commented on src/dstack/_internal/server/models.py:
```diff
@@ -230,3 +241,76 @@ class GatewayComputeModel(BaseModel):
+
+    offer: Mapped[str] = mapped_column(String(4000))
+
+    resource_spec_data: Mapped[Optional[str]] = mapped_column(String(4000))
```
Unclear why having optional resource_spec_data. Remove if not needed.

TheBits replied:
Fixed.

r4victor commented on src/dstack/_internal/server/models.py:
```diff
@@ -230,3 +241,76 @@ class GatewayComputeModel(BaseModel):
+    job: Mapped[Optional["JobModel"]] = relationship(back_populates="instance")
+    last_job_processed_at: Mapped[Optional[datetime]] = mapped_column(DateTime)
+
+    # + # job_id: Optional[FK] (current job)
```
What's this comment is about? Leave readable TODO

TheBits replied:
Removed.

r4victor commented on src/dstack/_internal/cli/commands/run.py:
```diff
@@ -89,6 +114,31 @@ def _command(self, args: argparse.Namespace):
+        termination_policy_idle = 5 * 60  # 5 minutes by default
+        termination_policy = TerminationPolicy.DESTROY_AFTER_IDLE
+
+        if args.idle_duration is not None:
+            try:
+                termination_policy_idle = int(args.idle_duration)
+            except ValueError:
+                termination_policy_idle = 60 * parse_pretty_duration(args.idle_duration)
+
+        creation_policy = (
+            CreationPolicy.REUSE if args.creation_policy_reuse else CreationPolicy.REUSE_OR_CREATE
+        )
+
+        if creation_policy == CreationPolicy.REUSE and termination_policy_idle is not None:
+            console.print(
+                "[warning]If the flag --reuse is set, the argument --idle-duration will be skipped[/]"
+            )
+            termination_policy = TerminationPolicy.DONT_DESTROY
+
+        if args.instance_name is not None and termination_policy_idle is not None:
+            console.print(
+                f"[warning]--idle-duration won't be applied to the instance {args.instance_name!r}[/]"
+            )
+            termination_policy = TerminationPolicy.DONT_DESTROY
+
```
The pool-specific argument parsing should be refactored. See how other flags are parsed.

r4victor commented on src/dstack/_internal/server/models.py:
```diff
@@ -230,3 +241,76 @@ class GatewayComputeModel(BaseModel):
+    pool: Mapped["PoolModel"] = relationship(back_populates="instances", single_parent=True)
+
+    status: Mapped[InstanceStatus] = mapped_column(Enum(InstanceStatus))
+    status_message: Mapped[Optional[str]] = mapped_column(String(50))
```
Not used? Remove

TheBits replied:
Removed.

r4victor commented on .pre-commit-config.yaml:
```diff
@@ -1,7 +1,16 @@
+      args: ['--strict', '--follow-imports=skip', '--ignore-missing-imports', '--python-version=3.8']
+      files: '.*pools?\.py'
+      exclude: 'versions|src/tests'
+      additional_dependencies: [types-PyYAML, types-requests, pydantic<2, sqlalchemy]
```
I insist we should not make such an ad-hoc type checking a part of the CI. Type checking should be discussed in a separate issue. Remove from this PR.

TheBits replied:
I removed mypy from pull request

Egor-S commented on runner/internal/shim/docker.go:
```diff
@@ -19,109 +19,159 @@ import (
+
+	log.Printf("Container finished successfully, id=%s\n", containerID)
+
+	d.state = Pending
```
Is it the correct status? We may want to distinguish "no runs" and failed/done runs.

TheBits replied:
That's correct. We need "no runs" status. The runner can take new tasks when the status is "pending".

Egor-S commented on runner/internal/shim/api/http.go:
```diff
@@ -1,57 +1,53 @@
-		return nil, err
-	}
-	s.registryAuth <- base64.URLEncoding.EncodeToString(encodedConfig)
+	go func(taskParams shim.DockerImageConfig) {
```
We should have a way to kill a running container without interrupting the instance. I wouldn't rely on the runner in this case.

TheBits replied:
I'm going to make new issue for this.

TheBits replied:
#895

r4victor commented on src/dstack/_internal/server/routers/pools.py:
```diff
@@ -0,0 +1,119 @@
+    await pools.remove_instance(session, project_model, body.pool_name, body.instance_name)
+
+
+@router.post("/set-default")  # type: ignore[misc]
```
/set_default

TheBits replied:
Fixed

r4victor committed:
Fix currency missing in dstack pool show PRICE

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -39,10 +64,20 @@ def run_job(
+        instance_config: InstanceConfiguration,
+    ) -> LaunchedInstanceInfo:
+        pass
+
```
ProjectModel and UserModel cannot be a part of the backend interface. `core` modules should NOT import from `server`.

TheBits replied:
Thanks for the comment! Fixed

r4victor commented on src/dstack/_internal/server/services/runs.py:
```diff
@@ -118,35 +152,203 @@ async def get_run(
+    instance_config = InstanceConfiguration(
+        instance_name=instance_name,
+        pool_name=pool_name,
+        ssh_keys=[ssh_key],
```
User's ssh key is not added, so the CLI won't be able to connect to the instance.

Commit:
Fix ssh keys. Fix review

Commit:
Merge remote-tracking branch 'origin/issue_790_dstack_pool' into issue_790_dstack_pool

Commit:
Merge remote-tracking branch 'origin/master' into issue_790_dstack_pool

Egor-S committed:
Do not require repo in `dstack add`

Egor-S committed:
Show provisioned instance in `dstack pool add`

Egor-S committed:
Validate `pool add` resources args

Egor-S committed:
Print pool name in run plan

Egor-S committed:
Add TODOs

Egor-S committed:
Add InstanceAvailability for pool instances

Commit:
Timout for runner download

Commit:
Remove duplicate code. Fix termination_idle_time. Small review fixes

Commit:
Merge remote-tracking branch 'origin/master' into issue_790_dstack_pool

Commit:
Merge remote-tracking branch 'origin/issue_790_dstack_pool-improve-cli' into issue_790_dstack_pool

r4victor commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -2,25 +2,44 @@
+    def get_public_keys(self) -> List[str]:
+        return [ssh_key.public.strip() for ssh_key in self.ssh_keys]
+
+
```
All these models should be in core/models.

TheBits replied:
Fixed

r4victor commented on src/dstack/_internal/server/background/tasks/process_pools.py:
```diff
@@ -0,0 +1,181 @@
+    resp = shim_client.healthcheck()
+    if resp is None:
+        return False  # shim is not available yet
+    return bool(resp.service == "dstack-shim")
```
Redundant bool() conversion

TheBits replied:
Fixed.

Commit:
Always load job

Commit:
Merge remote-tracking branch 'origin/master' into issue_790_dstack_pool

Egor-S committed:
TODO and small fix

Commit:
Fix job_name

Egor-S committed:
`dstack show` now works with no pool provided

Egor-S committed:
Improve dstack pool show formatting

Commit:
Merge remote-tracking branch 'origin/issue_790_dstack_pool-default-pool' into issue_790_dstack_pool

Egor-S committed:
Ask confirmation on `dstack pool remove`

r4victor committed:
Fix done jobs being aborted

Fixes #885. The problem was that after reading the logs, the run status
wasn't done yet at the moment the CLI aborted it.
Added waiting time for run to finish.

Commit:
Remove mypy and comments

Commit:
Merge remote-tracking branch 'origin/issue_790_dstack_pool' into issue_790_dstack_pool

Commit:
Rename url pool/set-default to pool/set_default

Egor-S committed:
Replace instance.instance_id with instance.name

Egor-S committed:
dstack pool remove: name as positional argument

Commit:
fix review

Commit:
Merge remote-tracking branch 'origin/issue_790_dstack_pool-fix-names-mismatch' into issue_790_dstack_pool
